File: scripts/database_integration_scripts_V2/legacy/58_add_is_urban_settlement_to_cities.py
"""58 — Add is_urban_settlement to cities.

Mirrors `enhance_db/src/bin/58_add_is_urban_settlement_to_cities.rs`.

  Inputs : data/all_humans/entity_type_classification.json
              {qid: {urban_settlement: bool, ...}}
           cities.entity_type_ids   (pipe-joined Q-ids from step 57)
  Output : cities.is_urban_settlement (1 / 0 / NULL)

Usage
-----
    python3 58_add_is_urban_settlement_to_cities.py
    python3 58_add_is_urban_settlement_to_cities.py --full
"""

from __future__ import annotations

import json
import sqlite3
import tempfile
from pathlib import Path

from tqdm import tqdm

from common import (
    ALL_HUMANS_DIR,
    add_column_if_missing,
    insert_rows,
    load_json,
    log,
    open_db,
    parse_run_mode,
)

JSON_PATH = ALL_HUMANS_DIR / "entity_type_classification.json"


def run(conn: sqlite3.Connection, json_path: Path = JSON_PATH) -> int:
    log("[DB] 58: Add is_urban_settlement to cities...")
    raw = load_json(json_path)
    urban_ids: set[str] = {
        k for k, v in raw.items()
        if isinstance(v, dict) and v.get("urban_settlement") is True
    }
    log(f"[DB] {len(urban_ids)} urban P31 ids")

    add_column_if_missing(conn, "cities", "is_urban_settlement", "INTEGER")
    conn.execute("UPDATE cities SET is_urban_settlement = NULL")
    conn.commit()

    rows = list(conn.execute("SELECT id, entity_type_ids FROM cities"))
    cur = conn.cursor()
    urban = 0
    non_urban = 0
    null_n = 0
    for qid, etypes in tqdm(rows, desc="58", unit="city"):
        if not etypes:
            null_n += 1
            val = None
        else:
            val = 1 if any(t in urban_ids for t in etypes.split("|")) else 0
            if val == 1:
                urban += 1
            else:
                non_urban += 1
        cur.execute(
            "UPDATE cities SET is_urban_settlement = ? WHERE id = ?",
            (val, qid),
        )
    conn.commit()
    log(f"[DB] urban={urban} non_urban={non_urban} null={null_n}")
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_cities_is_urban_settlement "
        "ON cities(is_urban_settlement)"
    )
    conn.commit()
    return urban


def _sample_main() -> None:
    classification = {
        "Q515": {"urban_settlement": True},
        "Q174844": {"urban_settlement": True},
        "Q3957": {"urban_settlement": False},
    }
    with tempfile.TemporaryDirectory() as tmp:
        json_path = Path(tmp) / "classification.json"
        json_path.write_text(json.dumps(classification))
        db = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db) as seed:
            seed.execute(
                "CREATE TABLE cities (id TEXT PRIMARY KEY, name_en TEXT, "
                "entity_type_ids TEXT)"
            )
            insert_rows(seed, "cities", [
                {"id": "Q90", "name_en": "Paris", "entity_type_ids": "Q515|Q174844"},
                {"id": "Q1234", "name_en": "Village", "entity_type_ids": "Q3957"},
                {"id": "Q777", "name_en": "Mystery", "entity_type_ids": None},
            ])
        with open_db(db) as conn:
            run(conn, json_path=json_path)
            for r in conn.execute(
                "SELECT id, name_en, is_urban_settlement FROM cities"
            ):
                log(f"  {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()

File: scripts/legacy/database_integration_scripts/enhance_db/src/bin/58_add_is_urban_settlement_to_cities.rs
/// 58: Add `is_urban_settlement` column to `cities` based on the urban/not-urban
/// classification of Wikidata P31 classes produced by
/// `extraction_scripts/all_humans/36_classify_entity_types_urban.py`.
///
/// Rule: a city is an urban settlement (is_urban_settlement=1) if ANY of its
/// P31 ids is classified urban_settlement=true. Otherwise 0. NULL if the city
/// has no P31 data at all (should be rare).
use anyhow::{Context, Result};
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use serde_json::Value;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::Write;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const CLASSIFICATION_PATH: &str = "data/all_humans/entity_type_classification.json";
const TASK_LOG: &str = "task.log";

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

fn column_exists(conn: &Connection, table: &str, col: &str) -> Result<bool> {
    let mut stmt = conn.prepare(&format!("PRAGMA table_info({})", table))?;
    let names: Vec<String> = stmt
        .query_map([], |row| row.get::<_, String>(1))?
        .filter_map(|r| r.ok())
        .collect();
    Ok(names.iter().any(|n| n == col))
}

fn main() -> Result<()> {
    log("=== Step 58: Add is_urban_settlement to cities ===");

    let json = fs::read_to_string(CLASSIFICATION_PATH)
        .with_context(|| format!("reading {}", CLASSIFICATION_PATH))?;
    let raw: HashMap<String, Value> = serde_json::from_str(&json)?;
    log(&format!("[58] Loaded classification for {} P31 ids", raw.len()));

    // Build a set of P31 ids classified as urban.
    let urban_ids: HashSet<String> = raw
        .iter()
        .filter_map(|(k, v)| {
            if v.get("urban_settlement")
                .and_then(|x| x.as_bool())
                .unwrap_or(false)
            {
                Some(k.clone())
            } else {
                None
            }
        })
        .collect();
    log(&format!("[58] Urban P31 ids: {}", urban_ids.len()));

    let mut conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    if column_exists(&conn, "cities", "is_urban_settlement")? {
        log("[58] cities.is_urban_settlement already exists — clearing values");
        conn.execute("UPDATE cities SET is_urban_settlement = NULL", [])?;
    } else {
        conn.execute_batch("ALTER TABLE cities ADD COLUMN is_urban_settlement INTEGER;")?;
        log("[58] Added column cities.is_urban_settlement");
    }

    // Read all (id, entity_type_ids) rows. The entity_type_ids column is
    // pipe-separated P31 Q-ids (from step 57).
    let mut rows: Vec<(String, Option<String>)> = Vec::new();
    {
        let mut stmt = conn.prepare("SELECT id, entity_type_ids FROM cities")?;
        let it = stmt.query_map([], |r| {
            Ok((r.get::<_, String>(0)?, r.get::<_, Option<String>>(1)?))
        })?;
        for r in it {
            rows.push(r?);
        }
    }
    log(&format!("[58] Read {} cities", rows.len()));

    let tx = conn.transaction()?;
    let mut updated = 0usize;
    let mut urban_count = 0usize;
    let mut non_urban_count = 0usize;
    let mut null_count = 0usize;
    {
        let mut stmt =
            tx.prepare("UPDATE cities SET is_urban_settlement = ?1 WHERE id = ?2")?;
        let pb = ProgressBar::new(rows.len() as u64);
        pb.set_style(
            ProgressStyle::with_template(
                "  {msg} [{bar:40.cyan/blue}] {pos}/{len} ({percent}%) eta {eta}",
            )
            .unwrap()
            .progress_chars("=>-"),
        );
        pb.set_message("Updating cities");

        for (qid, etypes) in &rows {
            let val: Option<i64> = match etypes {
                None => None,
                Some(s) if s.is_empty() => None,
                Some(s) => {
                    let is_urban =
                        s.split('|').any(|tid| urban_ids.contains(tid));
                    Some(if is_urban { 1 } else { 0 })
                }
            };
            match val {
                Some(1) => urban_count += 1,
                Some(0) => non_urban_count += 1,
                _ => null_count += 1,
            }
            let u = stmt.execute(params![val, qid])?;
            updated += u as usize;
            pb.inc(1);
        }
        pb.finish_with_message("Updating cities done");
    }
    tx.commit()?;

    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_cities_is_urban_settlement ON cities(is_urban_settlement);",
    )?;

    log(&format!("[58] rows updated: {}", updated));
    log(&format!(
        "[58]   is_urban_settlement=1 : {} ({:.1}%)",
        urban_count,
        100.0 * urban_count as f64 / rows.len().max(1) as f64
    ));
    log(&format!(
        "[58]   is_urban_settlement=0 : {} ({:.1}%)",
        non_urban_count,
        100.0 * non_urban_count as f64 / rows.len().max(1) as f64
    ));
    log(&format!(
        "[58]   NULL                   : {} ({:.1}%)",
        null_count,
        100.0 * null_count as f64 / rows.len().max(1) as f64
    ));

    // Spot check
    let mut stmt = conn.prepare(
        "SELECT id, name_en, is_urban_settlement, entity_type FROM cities
         WHERE id IN ('Q100','Q30','Q771','Q60','Q90','Q1537')",
    )?;
    let spot: Vec<(String, Option<String>, Option<i64>, Option<String>)> = stmt
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?, r.get(3)?)))?
        .filter_map(|r| r.ok())
        .collect();
    log("[58] Spot check:");
    for (id, name, u, et) in spot {
        log(&format!(
            "[58]   {} {:<20} urban={} types={}",
            id,
            name.as_deref().unwrap_or("?"),
            match u {
                Some(1) => "YES",
                Some(0) => "no ",
                _ => "NULL",
            },
            et.as_deref().unwrap_or("-")
        ));
    }

    log("=== Step 58 complete ===");
    Ok(())
}
